internals/secrethub/audit.go

```diff
@@ -150,24 +329,180 @@ func (cmd *AuditCommand) iterAndAuditTable() (secrethub.AuditEventIterator, audi
 	return nil, nil, ErrNoValidRepoOrSecretPath
 }
 
+type pager interface {
+	io.WriteCloser
+	IsClosed() bool
+}
+
+// newPaginatedWriter runs the terminal pager configured in the OS environment
+// and returns a writer to its standard input.
+func newPaginatedWriter(outputWriter io.Writer) (pager, error) {
+	pager, err := pagerCommand()
+	if err != nil {
+		return nil, err
+	}
+
+	cmd := exec.Command(pager)
+
+	writer, err := cmd.StdinPipe()
+	if err != nil {
+		return nil, err
+	}
+
+	cmd.Stdout = outputWriter
+	cmd.Stderr = os.Stderr
+
+	err = cmd.Start()
+	if err != nil {
+		return nil, err
+	}
+	done := make(chan struct{}, 1)
+	go func() {
+		_ = cmd.Wait()
+		done <- struct{}{}
+	}()
+	return &paginatedWriter{writer: writer, cmd: cmd, done: done}, nil
+}
+
+type paginatedWriter struct {
+	writer io.WriteCloser
+	cmd    *exec.Cmd
+	done   <-chan struct{}
+	closed bool
+}
+
+func (p *paginatedWriter) Write(data []byte) (n int, err error) {
+	return p.writer.Write(data)
+}
+
+// Close closes the writer to the terminal pager and waits for the terminal pager to close.
+func (p *paginatedWriter) Close() error {
+	err := p.writer.Close()
+	if err != nil {
+		return err
+	}
+	if !p.closed {
+		<-p.done
+	}
+	return nil
+}
+
+// IsClosed checks if the terminal pager process has been stopped.
+func (p *paginatedWriter) IsClosed() bool {
+	if p.closed {
+		return true
+	}
+	select {
+	case <-p.done:
+		p.closed = true
```

Comment:
Calling `p.Close()` is not the solution here. The idea is that the user of the paginator can call `IsClosed()` to check if it can stop fetching the audit log entries, and call `Close()` if the program is supposed to end (there are no more entries or there is an error).

The `closed` field is used since we can only read from the `done` channel one time but the user should be able to call `IsClosed()` any number of times and receive correct return values. So once something is read from the channel, we know the pager is closed so we set the `closed` field and never check the channel again. This is the best solution I found for this issue and I think that modifying the `closed` field in `IsClosed()` is not a problem since, from an outside perspective, it does not alter the behavior of the `paginatedWriter` in any (subsequent) calls.